tools/autograd/gen_python_functions.py

```diff
@@ -656,180 +653,152 @@ def method_def(name, declarations, is_python_method, module):
+#     bitwise_and(Tensor input, Scalar other, *, Tensor out=None)
+#
+#   If they have different number of parameters then they are not ambiguous - but
+#   the difference on output param can be ignored as it's optional.
+#
+#     multiply(Tensor input, Tensor other, *, Tensor out=None)
+#     multiply(Tensor input, Scalar other)
+#
+#   Both positional args and keyword-only args are considered together.
+#
+#     subtract(Tensor other, *, Scalar alpha=1)
+#     subtract(Scalar other, Scalar alpha=1)
+#
+# A few ambiguous cases which it does NOT handle yet.
+#
+#   If there is any difference in other parameters besides the Tensor/Scalar
+#   difference, then they are not considered ambiguous by this method anymore.
+#   However, the difference could be too trivial to disambiguate.
+#
+#     foo(Tensor input, Scalar other, Scalar bar)
+#     foo(Tensor input, Tensor other, double bar)
+#
+#   If they are taking different number of parameters then they are not considered
+#   ambiguous anymore, even if the difference is only on optional kwargs.
+#
+#     foo(Scalar other, Scalar alpha=1)
+#     foo(Tensor other, *, Scalar alpha=1, Scalar beta=1)
+#
 
-    def dynamic_type(arg):
-        return arg['dynamic_type']
+def sort_overloads(
+    grouped_overloads: List[PythonSignatureGroup]
+) -> Tuple[PythonSignatureGroup, ...]:
 
-    def is_coord_smaller(arg1, arg2):
-        return dynamic_type(arg1) == 'Scalar' and arg2['dynamic_type'] == 'Tensor'
+    def is_arg_smaller(t1: Type, t2: Type) -> bool:
+        return str(t1) == 'Scalar' and str(t2) == 'Tensor'
 
-    def is_smaller(d1, d2):
-        """Returns True if d1 < d2 in the partial order."""
-        args1, args2 = d1['base']['arguments'], d2['base']['arguments']
+    def is_smaller(s1: PythonSignature, s2: PythonSignature) -> bool:
+        """Returns True if s1 < s2 in the partial order."""
+        args1, args2 = s1.arguments(skip_outputs=True), s2.arguments(skip_outputs=True)
         if len(args1) != len(args2):
             return False
-        any_smaller = any(is_coord_smaller(arg1, arg2) for arg1, arg2 in zip(args1, args2))
-        all_smaller_or_equal = all(dynamic_type(arg1) == dynamic_type(arg2) or
-                                   is_coord_smaller(arg1, arg2)
+        # TODO: should use some canonical form instead of 'str(arg.type)' - see comments
+        # above. The old codegen used the deprecated 'dynamic_type(arg.type)', which
+        # ignores the optional annotation, i.e. 'Scalar' and 'Scalar?'.
+        any_unequal = any(str(arg1.type) != str(arg2.type)
+                          for arg1, arg2 in zip(args1, args2))
+        all_smaller_or_equal = all(str(arg1.type) == str(arg2.type)
+                                   or is_arg_smaller(arg1.type, arg2.type)
                                    for arg1, arg2 in zip(args1, args2))
-        return any_smaller and all_smaller_or_equal
+        return any_unequal and all_smaller_or_equal
```

Comment:
I understand this is the fault of the original code, but I just want to remark: the use of "any" and "all" here obscures the essential logical property going on here: if you know how to compute `lte` and `eq`, you can compute `lt = lte && !eq`. The swapping between any/all is the result of pushing the negation inside of the all quantifier, but the quantifier doesn't really have anything to do with the logical computation going on here.

Reply:
Also, (once again, artifact of the old code) I'm a little perplexed as to why equality matters here, because I don't feel like there really should be any overloads in the list with exactly the same types, so given two distinct signatures they should always be unequal.